### PR TITLE
Add automatic function discovery

### PR DESCRIPTION
**Pull request by hiiambhanu**
Effort to address #2 
1. Added a `--discover` flag which is a `Boolean`, indicating whether or not we want to use automatic functions and path discovery.
2. Added another argument called `indexFilePath` which points to the location of the `index.ts file` which exports all the functions (ususally the `src/index.ts` file). This arg is optional and is only used when used with `--discover` flag.
3. If the discover flag is used, it uses the `ts-morph` script to get function paths and rewrites the "functions section" of `.differspec.json` file.

### CODE REVIEW

File: src/discoverer/discoverer.spec.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { expect } from "chai";
+import { getFirebaseFunctionsAndPaths } from "./discoverer";
+
+describe("Function Discovery", function () {
+    this.timeout(0);
```
Comment (haroldadmin) [inline src/discoverer/discoverer.spec.ts:5]: Why is this needed?

File: src/discoverer/discoverer.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import { ExportDeclaration, ImportDeclaration, Project } from "ts-morph";
+
+const getPath = (declaration: ImportDeclaration | ExportDeclaration): string | undefined => {
+    const sourceFile = declaration.getModuleSpecifierSourceFile()?.getFilePath()?.toString();
+    if (sourceFile) return sourceFile;
+    return undefined;
```
Comment (haroldadmin) [inline src/discoverer/discoverer.ts]: This is the same as `return sourceFile`?

File: src/discoverer/discoverer.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            } else {
+                if (path) exportMap[func.getText()] = path;
+            }
+        }
+    }
+    return exportMap;
+};
+const project = new Project();
+project.addSourceFilesAtPaths("**/*.ts");
+export { getExports as getFirebaseFunctionsAndPaths };
```
Comment (haroldadmin) [inline src/discoverer/discoverer.ts]: Prefer adding whitespace before exports

**Review by haroldadmin — CHANGES_REQUESTED** [review]
There's a few nitpicks I pointed out in the review, but it looks great overall. Thanks for contributing this feature!

File: src/discoverer/discoverer.spec.ts (hunk 1 above)
Comment (hiiambhanu) [reply src/discoverer/discoverer.spec.ts:5]: The test was timing out on my machine (which was kinda low on resources when I was running it) so had to increase the timeout. Moreover, depending on the complexity of the project, function discovery may take time longer than 3000ms which, I think was the default timeout for this test.

**Review by haroldadmin — APPROVED** [review]
LGTM! Really appreciate that you updated the docs too.

**Pull request merged**